File: ios/VineTrack/App/SprayProgram/SprayProgramTerminology.swift
import Foundation

/// Canonical operator-facing wording for the Spray Program.
///
/// The internal contract still says `isTemplate` / `SprayJobTemplateService` /
/// `is_template`, and it should — that is what the row actually is in storage.
/// What an operator reads, however, is one vocabulary and one only:
///
/// - a reusable step in the vineyard's spray program is a **Program Step**
/// - the collection of those steps is the **Program**
/// - taking a step into the calculator is **Plan from Program**
/// - a spray with no program step behind it is a **One-off Spray**
/// - promoting an applied spray into the program is **Add to Program**
///
/// The word "template" is reserved for exactly one thing that genuinely is a
/// blank form: the import spreadsheet, which is surfaced as
/// `downloadImportCSV`. Keeping these strings in one place is what lets the
/// tests prove the old vocabulary is gone rather than merely absent from the
/// files someone remembered to check.
nonisolated enum SprayProgramTerminology {

    // MARK: - Nouns

    /// A single reusable step in the spray program.
    static let programStep = "Program Step"
    /// The vineyard's collection of reusable steps.
    static let program = "Program"
    /// Plural of `programStep`.
    static let programSteps = "Program Steps"

    // MARK: - Actions

    /// Entry point that starts a spray from an existing Program Step.
    static let planFromProgram = "Plan from Program"
    /// Entry point that starts a spray with nothing pre-filled.
    static let oneOffSpray = "One-off Spray"
    /// Creates a new reusable Program Step.
    static let addProgramStep = "Add Program Step"
    /// Promotes an applied spray record into the reusable program.
    static let addToProgram = "Add to Program"
    /// Supporting state shown once a record is already in the program.
    static let inProgram = "In Program"
    /// Banner for a Program Step that is shared with the admin portal.
    ///
    /// Deliberately "Synced", not "Managed". The Program is a shared vineyard
    /// resource — the portal and mobile edit the SAME `spray_jobs` row — so
    /// wording that described a locked, portal-only object now misleads. What
    /// the operator needs to know is that this is the shared Program Step, and
    /// that changing it here changes it everywhere.
    static let syncedWithAdminPortal = "Synced with Admin Portal"
    /// Alias kept so the banner has one canonical name at the call sites that
    /// describe the sync relationship rather than the step itself.
    static var portalSyncBanner: String { syncedWithAdminPortal }

    // MARK: - Import spreadsheet

    /// The blank import spreadsheet. Deliberately NOT called a template — that
    /// word now means a Program Step to the operator.
    static let downloadImportCSV = "Download Import CSV"
    static let importCSV = "Import CSV"

    // MARK: - Composed copy

    static let chooseProgramStepTitle = "Choose a Program Step"
    static let noProgramStepsAvailable = "No Program Steps Available"

    /// Subtitle for the `Plan from Program` entry point.
    /// - Parameter count: number of Program Steps currently available.
    static func planFromProgramSubtitle(count: Int) -> String {
        count == 0
            ? "No Program Steps yet — add one in Spray Program or the admin portal"
            : "Start from a saved Program Step (\(count) available)"
    }

    /// Every operator-facing string this type vends. Used by tests to prove the
    /// retired vocabulary cannot creep back in.
    static var allLabels: [String] {
        [
            programStep, program, programSteps,
            planFromProgram, oneOffSpray, addProgramStep, addToProgram,
            inProgram, syncedWithAdminPortal,
            downloadImportCSV, importCSV,
            chooseProgramStepTitle, noProgramStepsAvailable,
            planFromProgramSubtitle(count: 0),
            planFromProgramSubtitle(count: 3)
        ]
    }
}

File: android-vinetrack/app/src/main/java/com/rork/vinetrack/data/spray/SprayProgramTerminology.kt
package com.rork.vinetrack.data.spray

/**
 * The operator-facing Spray Program vocabulary — the Android mirror of the
 * iOS `SprayProgramTerminology`.
 *
 * The word "template" is banned from operator-facing copy: a reusable step in
 * the vineyard's spray program is a "Program Step", the collection is the
 * "Program", and a planned-but-not-started spray reads as "Upcoming". The
 * labels are pinned here so a test can prove both platforms use the same
 * words rather than five near-misses.
 */
object SprayProgramTerminology {
    const val PROGRAM: String = "Program"
    const val PROGRAM_STEP: String = "Program Step"
    const val ADD_PROGRAM_STEP: String = "Add Program Step"
    const val PLAN_FROM_PROGRAM: String = "Plan from Program"
    const val PLAN_SPRAY: String = "Plan Spray"
    const val ONE_OFF_SPRAY: String = "One-off Spray"
    const val UPCOMING: String = "Upcoming"
    const val DOWNLOAD_IMPORT_CSV: String = "Download Import CSV"
    const val LOG_SPRAY_RECORD: String = "Log Spray Record"
    const val EDIT_PROGRAM_STEP: String = "Edit Program Step"

    /**
     * Banner for a Program Step that is shared with the admin portal.
     * Deliberately "Synced", not "Managed": the Program is a shared vineyard
     * resource — the portal, iOS and Android edit the SAME `spray_jobs` row —
     * so wording that described a locked, portal-only object now misleads.
     */
    const val SYNCED_WITH_ADMIN_PORTAL: String = "Synced with Admin Portal"

    /** Every pinned label, for the vocabulary test. */
    val allLabels: List<String> = listOf(
        PROGRAM,
        PROGRAM_STEP,
        ADD_PROGRAM_STEP,
        PLAN_FROM_PROGRAM,
        PLAN_SPRAY,
        ONE_OFF_SPRAY,
        UPCOMING,
        DOWNLOAD_IMPORT_CSV,
        LOG_SPRAY_RECORD,
        EDIT_PROGRAM_STEP,
        SYNCED_WITH_ADMIN_PORTAL,
    )
}
